code/Pagetypes/FreshdeskPage.php

```diff
@@ -0,0 +1,147 @@
+            $filter['priority'] = $this->request->getVar('priority');
+        }
+
+        $tickets = $this->filterTickets($tickets, $filter);
+
+        return new PaginatedList($tickets, $this->request);
+    }
+
+    /**
+    * Filter tickets based on productID and status/priority filters
+    *
+    * @param array $tickets
+    * @return ArrayList $tickets
+    */
+    public function filterTickets($tickets, $filter)
+    {
+        $productID = false;
+        if (defined('FRESHDESK_PRODUCT_ID')) {
+            $productID = FRESHDESK_PRODUCT_ID;
+        }
+
+        $tickets = new ArrayList($tickets);
+
+        if ($productID) {
+            $tickets = $tickets->filter('product_id', $productID);
+        }
+
+        $filter = $this->validateFilter($filter);
+
+        if ($filter) {
+            foreach ($filter as $filterKey => $filterVal) {
+                $tickets = $tickets->filter($filterKey, $filterVal);
+            }
+        }
+
+        return $tickets;
+    }
+
+    /**
+     * Renders a form which can be used for filtering the tickets via template.
+     *
+     * @return Form $form
+     */
+    public function filterForm()
+    {
+        $statuses = array_combine(array_values($this->freshdesk->getStoredStatuses()), array_values($this->freshdesk->getStoredStatuses()));
+        $priorities = array_combine($this->freshdesk->getPriorities(), $this->freshdesk->getPriorities());
+
+        $currentStatus = 'Open';
+        if ($this->request->getVar('status') && null != $this->request->getVar('status')) {
+            $currentStatus = $this->request->getVar('status');
+        }
+
+        $currentPriority = 'any';
+        if ($this->request->getVar('priority') && null != $this->request->getVar('priority')) {
+            $currentPriority = $this->request->getVar('priority');
+        }
+
+        $fields = new FieldList(
+            DropdownField::create('status', 'status', $statuses, $currentStatus),
+            DropdownField::create('priority', 'priority', $priorities, $currentPriority)
+        );
+
+        $actions = new FieldList(
+            FormAction::create('Filter', 'Filter')
+        );
+
+        $form = new Form($this, '', $fields, $actions);
+        $form->setTemplate('FilterForm');
+        $form->setFormMethod('get');
+
+        return $form;
+    }
+
+    /**
+     * Ensures filter only contains allowed statuses and priorities. If unallowed it is unset.
+     *
+     * @param array $filter
+     *
+     * @return array $filter
+     */
+    private function validateFilter($filter)
+    {
+        if (isset($filter['priority'])) {
+            if (!in_array($filter['priority'], array_values($this->freshdesk->getPriorities())) || ($filter['priority'] == 'any')) {
+                unset($filter['priority']);
+            }
+        }
+
+        if (isset($filter['status'])) {
+            if (!in_array($filter['status'], array_values($this->freshdesk->getStatuses())) || ($filter['status'] == 'any')) {
```

Comment:
Having getStatuses and getStoredStatuses to choose from is slightly confusing - the caller of this function would preferably not be bothered with having a choice. Case in point here - avoiding a freshdesk API callout to "statuses" was unsuccessful because of this stray call.

I think appropriate approach here is to only provide `getStatuses` call, and then deal with the caching inside the API layer (or between here and the API layer). You will need a cache that persists between requests, here is some reading https://docs.silverstripe.org/en/3/developer_guides/performance/caching/#using-caches

The easy thing is to chuck this inside the FreshdeskAPI calls. Preferred but harder - figure out how to apply caching without having to change FreshdeskAPI class (ideally, we should be striving to separate concerns). For example SS templates come with the %cached% directive, which you could use if there. Or we could wrap FreshdeskAPI inside a CachedFreshdeskAPI. Or... I don't know :-)

Reply:
It'd probably be a good thing if all caching was done on the same level too - currently there is $_tickets inside the Controller (caching within the request) and there is also the need to cache getStatuses. I'd say, they should be on the same level of abstraction.